File: services/agent-mgmt/app/utils/metrics.ts
import { Registry, Counter, Histogram, collectDefaultMetrics } from 'prom-client';

export const registry = new Registry();

collectDefaultMetrics({ register: registry });

export const requestTotal = new Counter({
  name: 'request_total',
  help: 'Total number of HTTP requests',
  labelNames: ['method', 'path', 'status'] as const,
  registers: [registry],
});

export const requestDurationSeconds = new Histogram({
  name: 'request_duration_seconds',
  help: 'HTTP request duration in seconds',
  labelNames: ['method', 'path'] as const,
  buckets: [0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5],
  registers: [registry],
});

File: services/scoring/app/utils/metrics.py
"""
Prometheus metrics definitions for the scoring service.
Exposed on GET /metrics via the prometheus_client ASGI middleware.
"""
from prometheus_client import Counter, Gauge, Histogram

# ── Request-level metrics ──────────────────────────────────────────────────────
REQUEST_TOTAL = Counter(
    "request_total",
    "Total HTTP requests received",
    ["method", "path", "status"],
)

REQUEST_DURATION = Histogram(
    "request_duration_seconds",
    "HTTP request latency in seconds",
    ["method", "path"],
)

# ── Scoring-specific metrics ──────────────────────────────────────────────────
SCORE_VALUE = Gauge(
    "score_value",
    "Most-recent score computed for an agent in a context type",
    ["agent_id", "context_type"],
)

BAND_WIDTH = Gauge(
    "band_width",
    "Current adaptive band width (high - low) for an agent + context type",
    ["agent_id", "context_type"],
)
